File: tests/sdk/go/securityhub_findings_test.go
package sdk_test

import (
	"context"
	"fmt"
	"testing"
	"time"

	"github.com/aws/aws-sdk-go-v2/service/sts"
)

func TestSecurityHubBatchImportAndGetFindings(t *testing.T) {
	requireReady(t)
	prefix := uniquePrefix(t)
	region := envOr("AWS_DEFAULT_REGION", "us-east-1")
	findingID := fmt.Sprintf("sdk-sh-%s", prefix)
	generatorID := "noctaxris-sdk-lab"
	now := time.Now().UTC().Format(time.RFC3339)

	stsClient := newSTS(t, loadAWSConfig(t))
	caller, err := stsClient.GetCallerIdentity(context.Background(), &sts.GetCallerIdentityInput{})
	if err != nil {
		t.Fatalf("GetCallerIdentity: %v", err)
	}
	accountID := ""
	if caller.Account != nil {
		accountID = *caller.Account
	}
	if accountID == "" {
		t.Fatal("GetCallerIdentity missing Account")
	}
	productArn := fmt.Sprintf("arn:aws:securityhub:%s:%s:product/%s/default", region, accountID, accountID)

	importStatus, importBody, importParsed := signedJSONTarget(t, "securityhub", "SecurityHub.BatchImportFindings", map[string]any{
		"Findings": []map[string]any{{
			"SchemaVersion": "2018-10-08",
			"Id":            findingID,
			"ProductArn":    productArn,
			"GeneratorId":   generatorID,
			"AwsAccountId":  accountID,
			"Types":         []string{"Software and Configuration Checks/Vulnerabilities/CVE"},
			"CreatedAt":     now,
			"UpdatedAt":     now,
			"Severity":      map[string]any{"Label": "HIGH"},
			"Title":         "sdk lab finding",
			"Description":   "security hub sdk import",
			"Resources": []map[string]any{{
				"Type": "AwsS3Bucket",
				"Id":   "arn:aws:s3:::sdk-sh-lab",
			}},
		}},
	})
	if importStatus != 200 {
		t.Fatalf("BatchImportFindings status=%d body=%s", importStatus, importBody)
	}
	successCount, _ := importParsed["SuccessCount"].(float64)
	failedCount, _ := importParsed["FailedCount"].(float64)
	if successCount != 1 || failedCount != 0 {
		t.Fatalf("BatchImportFindings parsed=%v", importParsed)
	}

	getStatus, getBody, getParsed := signedJSONTarget(t, "securityhub", "SecurityHub.GetFindings", map[string]any{
		"GeneratorId":   generatorID,
		"SeverityLabel": "HIGH",
		"ResourceType":  "AwsS3Bucket",
	})
	if getStatus != 200 {
		t.Fatalf("GetFindings status=%d body=%s", getStatus, getBody)
	}
	findings, _ := getParsed["Findings"].([]any)
	found := false
	for _, item := range findings {
		f, _ := item.(map[string]any)
		if f == nil {
			continue
		}
		id, _ := f["Id"].(string)
		title, _ := f["Title"].(string)
		if id == findingID || title == "sdk lab finding" {
			found = true
			break
		}
	}
	if !found {
		t.Fatalf("GetFindings missing imported finding: %s body=%s", findingID, getBody)
	}
}

File: tests/sdk/nodejs/test/securityhub-findings.test.mjs
import { test } from "node:test";
import assert from "node:assert/strict";
import { GetCallerIdentityCommand } from "@aws-sdk/client-sts";
import {
  newSTS,
  requireReady,
  signedJsonTarget,
  uniquePrefix,
} from "../lib/helpers.mjs";

test("Security Hub BatchImportFindings then GetFindings", async (t) => {
  if (!(await requireReady(t))) return;

  const prefix = uniquePrefix();
  const region = process.env.AWS_DEFAULT_REGION || "us-east-1";
  const findingId = `sdk-sh-${prefix}`;
  const generatorId = "noctaxris-sdk-lab";
  const now = new Date().toISOString();

  const sts = newSTS();
  const caller = await sts.send(new GetCallerIdentityCommand({}));
  const accountId = caller.Account;
  assert.ok(accountId, "GetCallerIdentity missing Account");
  const productArn = `arn:aws:securityhub:${region}:${accountId}:product/${accountId}/default`;

  const imported = await signedJsonTarget(
    "securityhub",
    "SecurityHub.BatchImportFindings",
    {
      Findings: [
        {
          SchemaVersion: "2018-10-08",
          Id: findingId,
          ProductArn: productArn,
          GeneratorId: generatorId,
          AwsAccountId: accountId,
          Types: [
            "Software and Configuration Checks/Vulnerabilities/CVE",
          ],
          CreatedAt: now,
          UpdatedAt: now,
          Severity: { Label: "HIGH" },
          Title: "sdk lab finding",
          Description: "security hub sdk import",
          Resources: [
            {
              Type: "AwsS3Bucket",
              Id: "arn:aws:s3:::sdk-sh-lab",
            },
          ],
        },
      ],
    },
  );
  assert.equal(
    imported.status,
    200,
    `BatchImportFindings status=${imported.status} body=${imported.body}`,
  );
  assert.equal(imported.json?.SuccessCount, 1);
  assert.equal(imported.json?.FailedCount, 0);

  const got = await signedJsonTarget("securityhub", "SecurityHub.GetFindings", {
    GeneratorId: generatorId,
    SeverityLabel: "HIGH",
    ResourceType: "AwsS3Bucket",
  });
  assert.equal(
    got.status,
    200,
    `GetFindings status=${got.status} body=${got.body}`,
  );
  const findings = got.json?.Findings || [];
  const found = findings.some(
    (f) => f?.Id === findingId || f?.Title === "sdk lab finding",
  );
  assert.ok(found, `GetFindings missing imported finding: ${got.body}`);
});
